engine/include/cubos/engine/ui/canvas/draw_list.hpp

```diff
@@ -0,0 +1,74 @@
+#include <vector>
+
+#include <cubos/core/gl/render_device.hpp>
+
+#include <cubos/engine/ui/canvas/draw_command.hpp>
+#include <cubos/engine/ui/canvas/draw_type.hpp>
+
+namespace cubos::engine
+{
+    /// @brief Describes a draw commands and how it fits into the draw list carrying it.
+    /// @ingroup ui-canvas-plugin
+    struct UIDrawListEntry
+    {
+        UIDrawType& type;
+        UIDrawCommand command;
+        size_t offset;
+    };
+
+    /// @brief Holds a collections of draw commands and their data.
+    /// @ingroup ui-canvas-plugin
+    class UIDrawList
+    {
+    public:
+        /// @brief Constructs a draw list.
+        UIDrawList();
+
+        /// @brief Move constructs.
+        /// @param other Other draw list.
+        UIDrawList(UIDrawList&& other) noexcept
+            : entries(other.entries)
+            , perElementData(other.perElementData)
+        {
+            other.perElementData = nullptr;
+        }
+
+        /// @brief Copy constructs.
+        /// @param other Other draw list.
+        UIDrawList(const UIDrawList& other) noexcept
+            : entries(other.entries)
+        {
+            perElementData = operator new(sizeof(char) * 1024);
+            std::memcpy(perElementData, other.perElementData, sizeof(char) * 1024);
+        }
+        ~UIDrawList()
+        {
+            operator delete(perElementData);
+        }
+
+        /// @brief Adds entry to draw list.
+        /// @param entry Entry to add.
+        /// @param data Pointer to per element data.
+        void pushEntry(UIDrawListEntry entry, const void* data);
+
+        /// @brief Creates new draw command and adds it to draw list.
+        /// @param type Type of the command.
+        /// @param vertexArray Vertex array handle.
+        /// @param vertexOffset Index of the first vertex to be drawn.
+        /// @param vertexCount Number of vertexes to draw.
+        /// @param data Pointer to per element data.
+        /// @return The created command.
```

Comment:
Why do you return a reference? This is dangerous, as the reference may become invalid the next time the method is called (due to std::vector moving the memory around).
If you just want the user to be able to customize the command further, I would return a builder with helper methods. This makes it more clear that its supposed to be used immediately after and not kept around in a variable.